### PR TITLE
Updates demo script to execute with bundled jdk

### PR DESCRIPTION
**Pull request by DarshitChanpura**
### Description
This PR adds changes to the tool to allow execution with bundled jdk. It also refactors usage of static variables.
It also addresses a small bug in `isStringAlreadyPresentFile` to modify from regex match to "startsWith". 

Tests are added as fast-follow in #3807 

### Check List
~- [ ] New functionality includes testing~ <- fast-follow via #3807 
~- [ ] New functionality has been documented~
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/security/tools/democonfig/Installer.java [hunk 1]
```diff
@@ -205,23 +205,28 @@ static void setOpenSearchVariables() {
         OPENSEARCH_LIB_PATH = BASE_DIR + "lib" + File.separator;
         OPENSEARCH_INSTALL_TYPE = determineInstallType();
 
+        boolean shouldExit = false;
         if (!(new File(OPENSEARCH_CONF_FILE).exists())) {
-            System.out.println("Unable to determine OpenSearch config directory. Quit.");
-            System.exit(-1);
+            System.out.println("Unable to determine OpenSearch config file. Quit.");
+            shouldExit = true;
```
Comment (peternied) [inline src/main/java/org/opensearch/security/tools/democonfig/Installer.java:211]: Nit: better control flow by using a collection to captures validation error, and then if you have any errors exiting

File: src/main/java/org/opensearch/security/tools/democonfig/Installer.java [hunk 2]
```diff
@@ -205,23 +205,28 @@ static void setOpenSearchVariables() {
         OPENSEARCH_LIB_PATH = BASE_DIR + "lib" + File.separator;
         OPENSEARCH_INSTALL_TYPE = determineInstallType();
 
+        boolean shouldExit = false;
         if (!(new File(OPENSEARCH_CONF_FILE).exists())) {
-            System.out.println("Unable to determine OpenSearch config directory. Quit.");
-            System.exit(-1);
+            System.out.println("Unable to determine OpenSearch config file. Quit.");
```
Comment (peternied) [inline src/main/java/org/opensearch/security/tools/democonfig/Installer.java:210]: Are we should be running the same checkstyle + spotless rules on these files as the rest of the codebase?

File: src/main/java/org/opensearch/security/tools/democonfig/Installer.java [hunk 3]
```diff
@@ -356,11 +361,37 @@ static void finishScriptExecution() {
...
             System.out.println(e.getMessage());
         }
     }
+
+    /**
+     * FOR TESTS ONLY
+     * Resets the class-level variables to their original values
+     * Is needed for tests since the variables are declared static
+     */
+    static void resetState() {
```
Comment (peternied) [inline src/main/java/org/opensearch/security/tools/democonfig/Installer.java:381]: If Installer was a class in the `static ... main()` method, you could just make an instance of the class, this would allow your tests to create fresh instances without any worry of static state management.

Maintaining this list as new stateful values are added and change behavior will be hard to check

File: src/test/java/org/opensearch/security/tools/democonfig/CertificateGeneratorTests.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            String fileContents = null;
+            try {
+                fileContents = new String(Files.readAllBytes(Path.of(certFilePath)));
+            } catch (Exception e) {
+                fail("Expected the test to pass.");
+            }
+
+            assertThat(fileContents.isEmpty(), not(true));
+            assertThat(fileContents, containsString("---BEGIN"));
+            assertThat(fileContents, containsString("---END"));
```
Comment (peternied) [inline src/test/java/org/opensearch/security/tools/democonfig/CertificateGeneratorTests.java:69]: Feels like we should load and verify items like; expiry time (10 years?), cert chain is consistent, etc ?

File: src/test/java/org/opensearch/security/tools/democonfig/InstallerTests.java [hunk 5]
```diff
@@ -0,0 +1,512 @@
...
+        assertEquals("/scriptDir", SCRIPT_DIR);
+        assertThat(assumeyes, is(true));
+        assertThat(initsecurity, is(true));
+        assertThat(cluster_mode, is(true));
+        assertEquals(0, skip_updates);
+        assertEquals(ExecutionEnvironment.TEST, environment);
+    }
+
+    @Test
+    public void testReadOptions_help() {
```
Comment (peternied) [inline src/test/java/org/opensearch/security/tools/democonfig/InstallerTests.java:115]: every time we add/change help text a unit test should break?  Can we verify this checking structured output instead of the process output?

**Review by peternied — COMMENTED** [review]
These tests look pretty brittle since they are handling output written to the console and files, I'd recommend breaking up how the parts of the tool work so you can have easier to maintain test cases.

Something to consider if the tool gets a new option, say <10 lines of code.  In the tests you would only really need two new test cases (pass/fail) and no unrelated tests would break.

File: src/main/java/org/opensearch/security/tools/democonfig/Installer.java (hunk 2 above)
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/tools/democonfig/Installer.java]: I see that you disagreed. Other files in tools package use the same format of using sysout to print information. To address this, we added an except to checkstyle [here](https://github.com/opensearch-project/security/blob/main/checkstyle/checkstyle.xml#L41)

**Comment by DarshitChanpura** [comment]
it is quite perplexing, that the test that are failing for windows only fail for 11 and 17 distributions of java. They work perfectly fine with 21.

I have reproduced these successfully on command prompt on windows machine. The interesting part is when I run the test on debugger in VS code on the same windows machine (using java 11), the test pass, but when I run the same on command prompt using java 11 the tests fail.

**Comment by DarshitChanpura** [comment]
the issue was with comparing characters when generating random password.

**Review by cwperks — COMMENTED** [review]
Took a first pass and left a few suggestions.

File: src/main/java/org/opensearch/security/tools/democonfig/Installer.java [hunk 6]
```diff
@@ -45,7 +45,7 @@ public class Installer {
 
     static ExecutionEnvironment environment = ExecutionEnvironment.DEMO;
 
-    static final String OS = System.getProperty("os.name") + " " + System.getProperty("os.version") + " " + System.getProperty("os.arch");
+    static String OS = System.getProperty("os.name") + " " + System.getProperty("os.version") + " " + System.getProperty("os.arch");
```
[the remark this answers is not in the document]
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/tools/democonfig/Installer.java:48]: yes, for testing. I modify these for unit tests (testing output when OS is Windows, something else, etc) and hence is required to be non-final

File: src/test/java/org/opensearch/security/sanity/tests/SecurityRestTestCase.java [hunk 7]
```diff
@@ -45,6 +45,8 @@ public class SecurityRestTestCase extends OpenSearchRestTestCase {
 
     private static final String CERT_FILE_DIRECTORY = "sanity-tests/";
 
+    static boolean useAdminAsPassword = false;
```
[the remark this answers is not in the document]
Comment (DarshitChanpura) [reply src/test/java/org/opensearch/security/sanity/tests/SecurityRestTestCase.java:48]: yes, It is required by buildclient to establish what password to be used for building an admin user client. Since build client is invoked before test-setup, the only way to access it is via static variable (modified via @BeforeClass)

File: src/main/java/org/opensearch/security/tools/democonfig/Installer.java (hunk 3 above)
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/tools/democonfig/Installer.java]: okie. I have rewrote the tool to follow singleton pattern.

File: src/main/java/org/opensearch/security/tools/democonfig/Installer.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/tools/democonfig/Installer.java]: The tool now follows singleton pattern

File: src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java [hunk 8]
```diff
@@ -193,46 +194,68 @@ static void writeSecurityConfigToOpenSearchYML() {
...
-        securityConfigLines.append("plugins.security.enable_snapshot_restore_privilege:  true\n");
-        securityConfigLines.append("plugins.security.check_snapshot_restore_write_privileges:  true\n");
-        securityConfigLines.append("plugins.security.restapi.roles_enabled:  [\"all_access\", \"security_rest_api_access\"]\n");
-
-        securityConfigLines.append("plugins.security.system_indices.enabled: true\n");
-        securityConfigLines.append("plugins.security.system_indices.indices: [").append(SYSTEM_INDICES).append("]\n");
+        securityConfigLines.append(
+            "plugins.security.authcz.admin_dn:"
+                + System.lineSeparator()
+                + "  - CN=kirk,OU=client,O=client,L=test, C=de"
```
[the remark this answers is not in the document]
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java]: yep..updated

**Comment by peternied** [comment]
> These tests look pretty brittle since they are handling output written to the console and files, I'd recommend breaking up how the parts of the tool work so you can have easier to maintain test cases.
> 
> Something to consider if the tool gets a new option, say <10 lines of code. In the tests you would only really need two new test cases (pass/fail) and no unrelated tests would break.

This comment isn't addressed.

File: src/test/java/org/opensearch/security/sanity/tests/SingleClusterSanityIT.java [hunk 9]
```diff
@@ -34,6 +38,13 @@ public void testSecurityPluginInstallation() throws Exception {
         verifyPluginInstallationOnAllNodes();
     }
 
+    @Test
+    public void testAdminCredentials_customStrongPassword_shouldSucceed() throws Exception {
```
[the remark this answers is not in the document]
Comment (DarshitChanpura) [reply src/test/java/org/opensearch/security/sanity/tests/SingleClusterSanityIT.java:42]: performRequest() requires an exception to be handled or thrown. I chose to not handle it since the test should fail if performRequest didn't work as expected.

File: src/test/java/org/opensearch/security/sanity/tests/InvalidAdminPasswordIT.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+    @BeforeClass
+    public static void setUpInvalidAdminCredentials() {
+        useAdminAsPassword = true;
+    }
+
+    @AfterClass
+    public static void restoreValidAdminCredentials() {
+        useAdminAsPassword = false;
+    }
```
Comment (derek-ho) [inline src/test/java/org/opensearch/security/sanity/tests/InvalidAdminPasswordIT.java:36]: Why is this necessary? Does the useAdminAsPassword carry over to other files?

Comment (DarshitChanpura) [reply src/test/java/org/opensearch/security/sanity/tests/InvalidAdminPasswordIT.java:36]: this is to test that the request is not successful when the password is set to `admin`.

File: src/main/java/org/opensearch/security/tools/democonfig/Installer.java [hunk 11]
```diff
@@ -356,11 +383,19 @@ static void finishScriptExecution() {
...
+                    + "."
+            );
             System.out.println("### (Ignore the SSL certificate warning because we installed self-signed demo certificates)");
 
         } catch (Exception e) {
             System.out.println(e.getMessage());
         }
     }
+
+    public static void resetInstance() {
```
Comment (cwperks) [inline src/main/java/org/opensearch/security/tools/democonfig/Installer.java:398]: If this isn't used anywhere can it be removed?

File: src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java [hunk 12]
```diff
@@ -151,15 +187,15 @@ static void writePasswordToInternalUsersFile(String adminPassword, String intern
     /**
      * Checks if security plugin is already configured. If so, the script execution will not continue.
      */
-    static void checkIfSecurityPluginIsAlreadyConfigured() {
+    void checkIfSecurityPluginIsAlreadyConfigured() {
         // Check if the configuration file contains the 'plugins.security' string
-        if (OPENSEARCH_CONF_FILE != null && new File(OPENSEARCH_CONF_FILE).exists()) {
-            try (BufferedReader br = new BufferedReader(new FileReader(OPENSEARCH_CONF_FILE, StandardCharsets.UTF_8))) {
+        if (this.installer.OPENSEARCH_CONF_FILE != null && new File(this.installer.OPENSEARCH_CONF_FILE).exists()) {
```
Comment (cwperks) [inline src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java:192]: nit: is `this` required everywhere here?

Would it make sense to move this method to the Installer or consider making getters for the values being accessed here?

File: src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java [hunk 13]
```diff
@@ -175,66 +211,75 @@ static void checkIfSecurityPluginIsAlreadyConfigured() {
...
+        String configFooter = "######## End OpenSearch Security Demo Configuration ########" + System.lineSeparator();
+
+        Map<String, Object> securityConfigAsMap = buildSecurityConfigMap();
+
+        try (FileWriter writer = new FileWriter(this.installer.OPENSEARCH_CONF_FILE, StandardCharsets.UTF_8, true)) {
+            writer.write(configHeader);
+            Yaml yaml = new Yaml();
+            DumperOptions options = new DumperOptions();
+            options.setDefaultFlowStyle(DumperOptions.FlowStyle.BLOCK);
+            String yamlString = yaml.dump(securityConfigAsMap);
```
Comment (cwperks) [inline src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java:230]: Nice!

File: src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java [hunk 14]
```diff
@@ -151,15 +187,15 @@ static void writePasswordToInternalUsersFile(String adminPassword, String intern
...
-            try (BufferedReader br = new BufferedReader(new FileReader(OPENSEARCH_CONF_FILE, StandardCharsets.UTF_8))) {
+        if (this.installer.OPENSEARCH_CONF_FILE != null && new File(this.installer.OPENSEARCH_CONF_FILE).exists()) {
+            try (BufferedReader br = new BufferedReader(new FileReader(this.installer.OPENSEARCH_CONF_FILE, StandardCharsets.UTF_8))) {
                 String line;
                 while ((line = br.readLine()) != null) {
                     if (line.toLowerCase().contains("plugins.security")) {
-                        System.out.println(OPENSEARCH_CONF_FILE + " seems to be already configured for Security. Quit.");
-                        System.exit(skip_updates);
+                        System.out.println(this.installer.OPENSEARCH_CONF_FILE + " seems to be already configured for Security. Quit.");
+                        System.exit(this.installer.skip_updates);
```
Comment (cwperks) [inline src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java:198]: Is `skip_updated` ever changed from -1? Why have a variable?

**Review by cwperks — COMMENTED** [review]
Took another pass. Nice work getting SnakeYaml to write to opensearch.yml!

File: src/main/java/org/opensearch/security/tools/democonfig/Installer.java (hunk 11 above)
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/tools/democonfig/Installer.java:398]: It is used in tests

File: src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java (hunk 12 above)
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java:192]: We do need to keep the method here since we split the installer class to be slightly modular. This class handles security related config settings. Given that we are using singleton instance, installer might not need this since all instances of SecuritySettingsConfigurer will still refer to same Installer instance. Now that I think out loud, I'll remove this.

File: src/test/java/org/opensearch/security/tools/democonfig/InstallerTests.java (hunk 5 above)
Comment (DarshitChanpura) [reply src/test/java/org/opensearch/security/tools/democonfig/InstallerTests.java]: I'm moving this discussion to #3807 and resolving it here.

**Comment by DarshitChanpura** [comment]
Code Coverage will be enhanced in #3807

File: src/main/java/org/opensearch/security/tools/democonfig/Installer.java [hunk 15]
```diff
@@ -82,60 +142,45 @@ static void printScriptHeaders() {
...
+            CommandLine line = parser.parse(options, args);
+
+            if (line.hasOption("h")) {
+                showHelp();
+                return;
             }
+            assumeyes = line.hasOption("y");
+            initsecurity = line.hasOption("i");
+            cluster_mode = line.hasOption("c");
+            skip_updates = line.hasOption("s") ? 0 : -1;
```
Comment (scrawfor99) [inline src/main/java/org/opensearch/security/tools/democonfig/Installer.java:162]: Why is this the only boolean redefined into a number?

Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/tools/democonfig/Installer.java:162]: this is to defined exit code used later on line 196 in SecuritySettingsConfigurer.java

File: src/main/java/org/opensearch/security/tools/democonfig/Installer.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/tools/democonfig/Installer.java]: see comment below.

File: src/main/java/org/opensearch/security/tools/democonfig/Certificates.java
[the remark this answers is not in the document]
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/tools/democonfig/Certificates.java:1]: yes.

**Review by derek-ho — APPROVED** [review]

**Comment by derek-ho** [comment]
Ship it 🚢

File: src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java [hunk 16]
```diff
@@ -258,7 +300,7 @@ static boolean isNetworkHostAlreadyPresent(String filePath) {
      */
     static boolean isNodeMaxLocalStorageNodesAlreadyPresent(String filePath) {
         try {
-            String searchString = "^node.max_local_storage_nodes";
+            String searchString = "node.max_local_storage_nodes:";
             return isStringAlreadyPresentInFile(filePath, searchString);
```
Comment (cwperks) [inline src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java:304]: Is it possible to read the file in as YAML and check if the YAML object has the key you are searching for?

**Review by cwperks — APPROVED** [review]
This PR LGTM. 

One additional comment: Should the SecuritySettingsConfigurer get the values it needs in its constructor instead of the entire `installer` object?

File: src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java (hunk 16 above)
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java]: Thank you for catching this. I've updated it.

**Comment by DarshitChanpura** [comment]
> One additional comment: Should the SecuritySettingsConfigurer get the values it needs in its constructor instead of the entire installer object?

I started out with that but thought it might be better to pass just one reference `installer` rather than multiple local copies of variables inside Installer class. IMO, I'd prefer to keep it as is.

File: src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java [hunk 17]
```diff
@@ -258,30 +304,29 @@ static boolean isNetworkHostAlreadyPresent(String filePath) {
...
      */
-    static boolean isStringAlreadyPresentInFile(String filePath, String searchString) throws IOException {
-        try (BufferedReader reader = new BufferedReader(new FileReader(filePath, StandardCharsets.UTF_8))) {
-            String line;
-            while ((line = reader.readLine()) != null) {
-                if (line.matches(searchString)) {
-                    return true;
-                }
-            }
+    static boolean isStringAlreadyPresentInYMLFile(String filePath, String fieldName) throws IOException {
```
Comment (cwperks) [inline src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java:321]:
```suggestion
    static boolean isKeyPresentInYMLFile(String filePath, String fieldName) throws IOException {
```

**Review by cwperks — APPROVED** [review]

File: src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java (hunk 17 above)
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java]: done.

**Review by cwperks — APPROVED** [review]

**Review by scrawfor99 — APPROVED** [review]
Looks good to me! Thanks DC
